man/man3/acl_get_file.3
.TH ACL_GET_FILE 3
.SH NAME
acl_get_file, acl_set_file  \- get or set the ACL for a pathname
.SH SYNOPSIS
.B #include <acl/acl.h>
.PP
.B struct acl * acl_get_file(const char *path, acl_type_t type);
.br
.B "int acl_set_file(const char *path, acl_type_t type, struct acl *aclp);"
.SH DESCRIPTION
.I acl_get_file
returns a pointer to an allocated \f2struct acl\fP associated with the
pathname pointed to by \f2path\fP.  \f2type\fP determines whether the
default ACL (\f2type == ACL_TYPE_DEFAULT\fP) or access ACL (\f2type == ACL_TYPE_ACCESS\fP) is returned.  The default ACL is available only for directories.
If there is no default ACL associated with the specified directory, an
ACL containing zero entries is returned.
If
.B _POSIX_MAC
is in effect, then the process must have MAC read access to the object.
.PP
.I acl_set_file
sets the ACL of the specified pathname.  \f2type\fP indicates which ACL, 
default or access, is to be set.  Only directories can have a default ACL.
If \f2type\fP is not valid, the call fails.
The ACL is first checked for validity by
\f2acl_valid\fP(3).
The effective UID of the process must match the owner of the object
or the process must have appropriate privilege to set the access ACL
or the default ACL on
\f2path\fP.
If
.B _POSIX_CAP
is in effect, then the appropriate capability must include CAP_FOWNER.
In addition, if
.B _POSIX_MAC
is in effect, then the process must have MAC write access to the object.
.SH RETURN VALUES
.I acl_get_file
a pointer to an allocated \f2struct acl\fP if successful, NULL otherwise.
The storage should be freed with a call to \f2acl_free\fP with the returned
pointer as an argument when it is no longer needed.
.PP
.I acl_set_file
returns 0 if successful, -1 otherwise.
.SH ERRORS
.I acl_get_file:
.TP 16
EACCESS
Search permission is denied for a component
of the path prefix or the object exists and the process does not have
appropriate access rights.
.TP 16
EINVAL
\f2type\fP is not valid
.TP 16
ENAMETOLONG
The pathname or one of its components is too long.
.TP 16
ENOENT
The named object does not exist or \f2path\fP points to an empty string.
.TP 16
ENOMEM
allocation of the \f2struct acl\fP failed.
.TP 16
ENOSYS
ACL support is not available (not installed).
.TP 16
ENOTDIR
A component of the path prefix is not a directory.
.PP
.I acl_set_file:
.TP 16
EACCESS
Search permission is denied for a component
of the path prefix or the object exists and the process does not have
appropriate access rights.
.TP 16
EINVAL
\f2aclp\fP points to an invalid ACL.
\f2type\fP is not either ACL_TYPE_ACCESS, or ACL_TYPE_DEFAULT.
The ACL is too large (too many entries).
.TP 16
ENAMETOLONG
The pathname or one of its components is too long.
.TP 16
ENOENT
The named object does not exist or \f2path\fP points to an empty string.
.TP 16
ENOSPC
The file system is full or some other resource needed for the ACL storage
is not available.
.TP 16
ENOSYS
ACL support is not available (not installed).
.TP 16
ENOTDIR
A component of the path prefix is not a directory.
.TP 16
EPERM
The process does not have appropriate privilege to
perform the operation to set the ACL.
.TP 16
EROFS
This function requires modification of a file system which is currently
read-only.


doc/CHANGES
2.0.6 (08 April 2002)
* Fix a bug that causes setfacl not to report error messages if
  acl-set_file() fails (the bug was introduced in version 2.0.3).

2.0.5 (26 March 2002)
* Fix up acl_to_text compliance with POSIX draft.
* Add the new acl(5) manual page.
* Add a manual section 3 directory and Makefile.
* Add the libacl manual pages.
* Rename the existing acl_to_any_text function to __acl_to_any_text, and
  add a new acl_to_any_text function that only contains parameters that
  are actually useful.  Make acl_to_text and acl_to_any_text invoke
  __acl_to_any_text internally.  Update all uses of acl_to_any_text in
  getfacl/getfacl.c, setfacl/do_set.c, and chacl/chacl.c.
* Move the acl_size function form the LIBACL_CFILES to the POSIX_CFILES
  section (it is a POSIX function)
* Rearrange libacl routines so that several non-POSIX routines which
  were being exported are no longer exported (eg. acl_entry_to_any_str,
  acl_get_file_mode, acl_get_fd_mode, acl_set_file_mode, acl_set_fd_mode).

2.0.4 (12 March 2002)
* Man page updates from Andreas
* Test script updates from Andreas

2.0.3 (05 March 2002)
* Clean up the --default option handling in setfacl. The old
  workarounds caused a bug for unusual input.
* Changes to the --test output format setfacl generates: ACLs that
  are not changed are now displayed as `*'.
* Fix a bug in setfacl/sequence.c:seq_delete_cmd().
* Minor changes to test scripts

2.0.2 (01 March 2002)
* Apply several patches from Andreas, namely:
* man page fixes
* libacl code reformatting
* acl_from_text errno handling

2.0.1 (27 February 2002)
* Apply Andreas Gruenbacher's diffs.
* Fix up chacl for deletion of access ACL to be in line with Andreas. 
* Incorporate the Debian packaging again.

2.0.0 (21 February 2002)
* Reworked to use the new official system call API.
* Sync up with the XFS project, the SGI folk now use this source.
* Jumped to version 2 to allow XFS users to upgrade
  (Rationale: the XFS ACL user tools were at version 1.1.X, and
  packaging tools like rpm, dpkg, etc. must be presented with a
  greater version number to allow an upgrade to proceed).
* Added the chacl command to ease migration for existing XFS users,
  and for compatibility with IRIX.
* Added a flag to allow acl_print to produce a single-line ACL, in
  addition to the multi-line format.
* Extended attribute documentation has moved into the extended
  attribute package from SGI ("attr"), this ACL package now deals
  exclusively with ACLs.
* acl_from_text sometimes did not set errno when failing.
* Moved files and simplified #includes in libacl

0.7.22
* Create a shared version of libacl, instead of linking statically.
* Added missing error reporting to getfacl do_print() function.     

0.7.16
* Minor cleanups/fixes in the make files
* The test cases in test/src have been updated to match the
  corrected ACL entry syntax (see the 0.7.15 changes).

0.7.15
* acl_from_text() now accepts mask and other entries with either two
  or three fields (`other:rwx' or `other::rwx'). Posix 1003.1e specifies
  that these entries should have three entries, while under Solaris,
  they have only two fields.
* setfacl now parses mask and other entries with two or three entries.
* getfacl now produces mask and other entries with three fields
  (`mask::rwx' instead of the previous `mask:rwx').
* acl_get_fd(), acl_get_file(), acl_set_fd(), acl_set_file() now only
  fail with errno=ENOSYS if the underlying filesystem doesn't support
  ACLs. acl_get_fd() and acl_get_file() construct ACLs from the file
  mode permission bits if the kernel supports ACLs, but no ACL extended
  attributes are defined for the file (errno=ENOATTR).
* acl_to_text() wrongly separated ACL entires with ',' instead of '\n'.
* acl_from_text() only accepted three-letter permission strings
  (rwx, ---, rw-, etc.). Now also short strings (rw, -, etc.) are
  supported.

0.7.10
* acl_get_fd(), acl_get_file() now don't return basic ACL's anymore.
* acl_set_fd(), acl_set_file() fail on filesystems that don't support
  ACLs, even if the ACL to be set is a basice ACL.
* If msgfmt is not present on a system, the message catalog(s) are
  no longer being built.

0.7.8 (19 February 2001)
* Buffer overflow bug in ACL library (acl_to_any_text).
  (Reported by Charles Bertsch <CBertsch@microtest.com>)

0.7.2 (22 November 2000)
* Replace rman with groff for building HTML versions of the
  manual pages.
* Add descriptions of the permissions required for manipulating
  ACLs to the manual pages.
* Add long options to getfacl and setfacl, update the manual pages.
* Remove some small bugs from the src/showacl Perl script.
* showacl is now integrated in getfacl.
* Bug in acl_to_any_text() for ACLs with 0 entries fixed.
* Bug in acl_to_text(): Used '\n' as the ACL entry separator.
* German messages updated.

0.7.1 (23 October 2000)
* Basic GNU gettext support and a German translation of messages
  added.

0.7.0 (22 October 2000)
* Updates to the documentation
* Now an ACL_MASK entry is cloned from the ACL_GROUP_OBJ entry
  when needed also when the ACL_MASK entry is not re-calculated
  (-n option), so setfacl doesn't complain about an invalid ACL.

0.6.8 (11 October 2000)
* src/walk_tree.c was stat()ing the targets of symbolic links in
  WALK_PHYSICAL mode.
* Removed spurious definitions from src/gettext.h that caused
  problems with several versions of GNU gettext headers
* setfacl calls acl_delete_def_file() if removing the default ACL
  of a directory that doesn't have a default ACL. i(If the user lacks
  appropriate permissions, that no-operation fails.) Fixed.
* Tiny bug in src/showacl script (mask entry not applied to owning
  group entry)

0.6.5 (10 September 2000)
* Command line options revised for POSIX compatibility (now defaults to
  -H half-logical walk, added -L logical walk, -P physical walk).
* Funny post-order option no longer documented.

0.6.0 (24 March 2000)
* Permissions were not set to the union of the ACL_GROUP_OBJ and

libacl/acl_get_file.c
/*
  File: acl_get_file.c

  Copyright (C) 1999, 2000
  Andreas Gruenbacher, <a.gruenbacher@computer.org>

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU Library General Public
  License as published by the Free Software Foundation; either
  version 2 of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Library General Public License for more details.

  You should have received a copy of the GNU Library General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <attr/xattr.h>
#include <acl/libacl.h>
#include "libacl.h"
#include "__acl_from_xattr.h"

#include "byteorder.h"
#include "acl_ea.h"


/* 23.4.16 */
acl_t
acl_get_file(const char *path_p, acl_type_t type)
{
	const size_t size_guess = acl_ea_size(16);
	char *ext_acl_p = alloca(size_guess);
	const char *name;
	int retval;

	switch(type) {
		case ACL_TYPE_ACCESS:
			name = ACL_EA_ACCESS;
			break;
		case ACL_TYPE_DEFAULT:
			name = ACL_EA_DEFAULT;
			break;
		default:
			errno = EINVAL;
			return NULL;
	}

	if (!ext_acl_p)
		return NULL;
	retval = getxattr(path_p, name, ext_acl_p, size_guess);
	if (retval == -1 && errno == ERANGE) {
		retval = getxattr(path_p, ACL_EA_ACCESS, NULL, 0);
		if (retval > 0) {
			ext_acl_p = alloca(retval);
			if (!ext_acl_p)
				return NULL;
			retval = getxattr(path_p, name, ext_acl_p, retval);
		}
	}
	if (retval > 0) {
		acl_t acl = __acl_from_xattr(ext_acl_p, retval);
		return acl;
	} else if (retval == 0 || errno == ENOATTR) {
		struct stat st;

		if (stat(path_p, &st) != 0)
			return NULL;

		if (type == ACL_TYPE_DEFAULT) {
			if (S_ISDIR(st.st_mode))
				return acl_init(0);
			else {
				errno = EACCES;
				return NULL;
			}
		} else
			return acl_from_mode(st.st_mode);
	} else
		return NULL;
}

